### DISCUSSION TITLE
Architecture

### DISCUSSION THREAD

**Discussion by pmig** [glasskube/operator#4] [Technical]
# Glasskube Operator Architecture

## Frameworks & Programming language

There are multiple possible solutions to build a custom operator on top: 

 - Go based "Operator Framework": https://sdk.operatorframework.io/docs/building-operators/golang/
 - Java based "Java Operator SDK": https://javaoperatorsdk.io/
 
 We decided to use the Java Operator SDK in combination with Kotlin.
 
 ## Architecture

<img width="1552" alt="Operator" src="https://user-images.githubusercontent.com/3041752/188267923-3acac88e-c833-4c95-ad54-d214e254fe8d.png">

**Comment by waveywaves** [comment]
@pmig why java over golang ? was it because the apps that are being deployed with glass kube are java based and interoperability would be easier ? I wanted to contribute to the operator but the language is what stopped me from doing so for now. Looking forward to heading why Java was chosen over GlassKube. I have worked with fabric8io Kubernetes client in my time at Red Hat when I used to work on the jenkins openshift plugins so was quite surprised to see it be used here as well.

**Comment by pmig** [comment]
we initially started out with Java / Kotlin due to our previous experiences with these languages. The operator is in maintenance mode and superseded by our package manager https://github.com/glasskube/glasskube/

**Reply by waveywaves** [reply]
Using the languages you know definitely makes the development quicker and java is good as well if you have a dependency on certain java libraries. But thanks for letting me know I will checkout the other repo.
